### PR TITLE
Discussion task overview

### PR DESCRIPTION
**Pull request by polina-tsvilodub**
This section contains information about the motivation and theoretical background ported over from the last semester's report. I updated that only cosmetically. The sections about the completion metric and implementation process were reworked more fundamentally.

Therefore, for the word count I only consider the two sections `completion metrics` and `implementation process` where I did actual writing, resulting in: 
name | word count
------|------------
Polina Tsvilodub | 736

### CODE REVIEW

**Comment by EmmaZanoli** [comment]
> Therefore, for the word count I only consider the two sections `completion metrics` and `implementation process` where I did actual writing, resulting in: 365 + 371 = 736 (by Polina Tsvilodub)

Even if you are the sole contributor, I would suggest using the table contribution format used for other pull requests (see [this example](https://github.com/ALLUOS/ALLUOS.github.io/pull/54)) 😊

**Review by mheyen — APPROVED** [review]
Looks great! 
Only fixed some minor typos :)

Thanks!!

File: _posts/2021-03-02-discussion.md [hunk 1]
```diff
@@ -6,6 +6,150 @@ sections:
...
+
+As already noted, we wanted this task to enhance language production skills. Discussions are a staple of classroom teaching, not only in the context of second language learning. They provide opportunities to produce language in contextualized and purposeful ways: learners can practice applying form (e.g., grammar, vocabulary) and function (e.g., language used to clarify, explain, argue) to communicate and build ideas. Moreover, learners benefit from redundancy of ideas and the vocabulary associated with them: discussion allows for multiple opportunities to hear new concepts and content continuously explained, analyzed, and interpreted. Furthermore, discussions via written communication have been linked to the immersive experience of studying abroad (Oliva & Pollastrini, 1995), notoriously effective for language learning.
```
Comment (EmmaZanoli) [inline _posts/2021-03-02-discussion.md:52]:
```suggestion
As already noted, we wanted this task to enhance language production skills. Discussions are a staple of classroom teaching, not only in the context of second language learning. They provide opportunities to produce language in contextualized and purposeful ways: learners can practice applying form (e.g., grammar, vocabulary) and function (e.g., language used to clarify, explain, argue) to communicate and build ideas. Moreover, learners benefit from redundancy of ideas and the vocabulary associated with them: discussion allows for multiple opportunities to hear new concepts and content continuously explained, analyzed, and interpreted. Furthermore, discussions via written communication have been likened to the immersive experience of studying abroad (Oliva & Pollastrini, 1995), notoriously effective for language learning.
```

File: _posts/2021-03-02-discussion.md [hunk 2]
```diff
@@ -6,6 +6,150 @@ sections:
...
+
+Another additional feature can be corrections and helpful hints by the bot. Meanwhile, the bot should be similar as Barson (1993) describes the teacher in a similar classroom setting: only play a supportive role and help discreetly to identify errors while mostly appreciating the learner’s contributions.
+
+Furthermore, vocabulary hints and short descriptions of words which are possibly new to the learners would be a valuable addition. This way probably an even bigger vocabulary learning effect could be achieved.
+
+## Implementation process
+Below, the technical side of the discussion task reviewed. As part of our overall bot architecture, the implementation of the discussion task also relies on general modules like the `Task(ABC)` class and the room handler, described in the [implementation section]({{ "" | absolute_url }}/2021/04/02/implementation.html). Therefore, only the discussion-specific task handler is described here in detail.
```
Comment (EmmaZanoli) [inline _posts/2021-03-02-discussion.md:109]:
```suggestion
Here, we present the technical details of the implementation of this task. As part of our overall bot architecture, the implementation of the discussion task also relies on general modules like the `Task(ABC)` class and the room handler, described in the [implementation section]({{ "" | absolute_url }}/2021/04/02/implementation.html). Therefore, only the discussion-specific task handler is described here in detail.
```

File: _posts/2021-03-02-discussion.md [hunk 3]
```diff
@@ -6,6 +6,150 @@ sections:
...
+
+`Discussion(Task)` is a subclass of `Task(ABC)`. However, unlike the sentence correction and the vocabulary guessing tasks, the discussion task is not sequential in nature and, therefore, does not inherit from `SequentialTask`. Nevertheless, we were still able to reuse a significant part of the methods already implemented in said class.
+The task always starts off with the presentation of the text, which is retrieved from the database based on the users' proficiency. This is followed by three rounds of discussion.
+At the start of each round, a text-related question is sent and a timer is added to the job queue to determine when the round finishes, such that [adaptive graphics]({{ "" | absolute_url }}/2021/01/04/user/experience.html) can remind users to finish their discussion of the question on time.
```
Comment (EmmaZanoli) [inline _posts/2021-03-02-discussion.md:113]:
```suggestion
At the start of each round, a text-related question is sent and a timer is added to the job queue to determine when the round finishes, such that [visual reminders]({{ "" | absolute_url }}/2021/02/03/adaptive.html) can be sent at regular time intervals to remind users to finish their discussion of the question on time.
```

File: _posts/2021-03-02-discussion.md [hunk 4]
```diff
@@ -6,6 +6,150 @@ sections:
...
+
+Since we don't have to select for a specific user, all user input can be handled in the same `handle_question` state. Word counts and grammar scores for each user are stored in two dictionaries, which are required to satisfy the completion criteria of the task.
```
Comment (EmmaZanoli) [inline _posts/2021-03-02-discussion.md:115]:
```suggestion
Unlike other [tasks]({{ "" | absolute_url }}/#three), in this case we don't have to select a specific user, therefore all user input can be handled in the same `handle_question` state. Word counts and grammar scores for each user are stored in two dictionaries, which are required to satisfy the completion criteria of the task.
```

File: _posts/2021-03-02-discussion.md [hunk 5]
```diff
@@ -6,6 +6,150 @@ sections:
...
+After a message is sent, `update_word_counts(message, user)` and `grammar_classifier.classify(sentence)` are called in the `evaluate_user_input(update.message.from_user.username, update.message.text, task_no=1)` function. They are called on tokenized and preprocessed input in order to update those dictionaries with the information.
```
Comment (EmmaZanoli) [inline _posts/2021-03-02-discussion.md:116]:
```suggestion
After a message is sent, `update_word_counts(message, user)` and `grammar_classifier.classify(sentence)` are called in the `evaluate_user_input(update.message.from_user.username, update.message.text, task_no=1)` function. They are called on tokenized and preprocessed input in order to update those dictionaries with the relevant information.
```

File: _posts/2021-03-02-discussion.md [hunk 6]
```diff
@@ -6,6 +6,150 @@ sections:
...
+
+In the first and the second discussion rounds, additionally to evaluating user input, the `send_intermediate_feedback()` function is called.
+In the third and final discussion round, instead of intermediate feedback the final grammatical feedback is presented via calling `send_final_feedback()`.
```
Comment (EmmaZanoli) [inline _posts/2021-03-02-discussion.md:119]: I noticed that the mention of feedback comes up here for the first time, without a previous explanation. Even though there are several pointers to the GEC page, I'm afraid that someone who's only read this page so far might be confused by this. We might want to add a sentence or two about this earlier in the "completion metrics" section :)

**Review by EmmaZanoli — CHANGES_REQUESTED** [review]
I think you did a great job! I only made some minor suggestions here and there :)

**Pull request merged**